controllers/topology/machinedeployment_controller.go

```diff
@@ -0,0 +1,439 @@
+		return ctrl.Result{}, err
+	}
+
+	// Calculate which templates are still in use by MachineDeployments or MachineSets which are not in deleting.
+	templatesInUse, err := calculateTemplatesInUse(md, msList)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Delete unused templates of MachineSets which are in deleting.
+	if err := r.deleteMachineSetsTemplatesIfNecessary(ctx, templatesInUse, msList); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Delete unused templates of the MachineDeployment if it is in deleting.
+	if err := r.deleteMachineDeploymentTemplatesIfNecessary(ctx, templatesInUse, md); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+// getMachineSetsForDeployment returns a list of MachineSets associated with a MachineDeployment.
+func (r *MachineDeploymentReconciler) getMachineSetsForDeployment(ctx context.Context, md types.NamespacedName) ([]*clusterv1.MachineSet, error) {
+	// List MachineSets based on the MachineDeployment label.
+	msList := &clusterv1.MachineSetList{}
+	if err := r.Client.List(ctx, msList,
+		client.InNamespace(md.Namespace), client.MatchingLabels{clusterv1.MachineDeploymentLabelName: md.Name}); err != nil {
+		return nil, errors.Wrapf(err, "failed to list MachineSets for MachineDeployment/%s", md.Name)
+	}
+
+	// Copy the MachineSets to an array of MachineSet pointers, to avoid MachineSet copying later.
+	res := make([]*clusterv1.MachineSet, 0, len(msList.Items))
+	for i := range msList.Items {
+		res = append(res, &msList.Items[i])
+	}
+	return res, nil
+}
+
+func (r *MachineDeploymentReconciler) isClusterPaused(ctx context.Context, md *clusterv1.MachineDeployment, msList []*clusterv1.MachineSet) (bool, error) {
+	var cluster *clusterv1.Cluster
+	var err error
+
+	// If the MachineDeployment still exists, get the Cluster from the MachineDeployment.
+	if md != nil {
+		// Fetch the corresponding Cluster.
+		cluster, err = util.GetClusterByName(ctx, r.Client, md.ObjectMeta.Namespace, md.Spec.ClusterName)
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to check if Cluster is paused")
+		}
+	}
+
+	// Otherwise, get the Cluster from the first MachineSet, if possible.
+	if len(msList) > 0 {
+		cluster, err = util.GetClusterByName(ctx, r.Client, msList[0].ObjectMeta.Namespace, msList[0].Spec.ClusterName)
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to check if Cluster is paused")
+		}
+	}
+
+	if cluster == nil {
+		// This should never happen as at least one MachineDeployment or MachineSet should
+		// always exist during reconciliation.
+		return false, errors.Errorf("couldn't find a Cluster to check if it is paused")
+	}
+
+	return annotations.IsPaused(cluster, cluster), nil
+}
+
+// calculateTemplatesInUse returns all templates referenced in non-deleting MachineDeployment and MachineSets.
+func calculateTemplatesInUse(md *clusterv1.MachineDeployment, msList []*clusterv1.MachineSet) (map[string]bool, error) {
```

Comment:
Instead of calculating if the templates are in use, couldn't we just requeue and wait for the MachineSet to be deleted before proceeding?